frontend/src/employee-frontend/components/child-information/DailyServiceTimesSection.tsx

```diff
@@ -0,0 +1,706 @@
+// SPDX-FileCopyrightText: 2017-2021 City of Espoo
+//
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+import { Loading, Result } from 'lib-common/api'
+import { UUID } from 'lib-common/types'
+import CollapsibleSection from 'lib-components/molecules/CollapsibleSection'
+import { Label, P } from 'lib-components/typography'
+import React, { useContext, useState } from 'react'
+import { useEffect } from 'react'
+import { useTranslation } from '../../state/i18n'
+import { useRestApi } from 'lib-common/utils/useRestApi'
+import { SpinnerSegment } from 'lib-components/atoms/state/Spinner'
+import ErrorSegment from 'lib-components/atoms/state/ErrorSegment'
+import {
+  isRegular,
+  DailyServiceTimes,
+  isIrregular,
+  RegularDailyServiceTimes,
+  IrregularDailyServiceTimes
+} from '../../types/child'
+import {
+  deleteChildDailyServiceTimes,
+  getChildDailyServiceTimes,
+  putChildDailyServiceTimes
+} from '../../api/child/daily-service-times'
+import { faClock, faPen } from '../../../lib-icons'
+import styled from 'styled-components'
+import InlineButton from '../../../lib-components/atoms/buttons/InlineButton'
+import {
+  FixedSpaceColumn,
+  FixedSpaceRow
+} from 'lib-components/layout/flex-helpers'
+import Button from '../../../lib-components/atoms/buttons/Button'
+import { defaultMargins, Gap } from 'lib-components/white-space'
+import Radio from '../../../lib-components/atoms/form/Radio'
+import InputField from '../../../lib-components/atoms/form/InputField'
+import Checkbox from '../../../lib-components/atoms/form/Checkbox'
+import { UIContext } from '../../state/ui'
+import { RequireRole } from '../../utils/roles'
+
+interface Props {
+  id: UUID
+  open: boolean
+}
+
+interface TimeInputRange {
+  start: string
+  end: string
+}
+
+interface SelectableTimeInputRange extends TimeInputRange {
+  selected: boolean
+}
+
+interface FormData {
+  type: 'REGULAR' | 'IRREGULAR' | 'NOT_SET'
+  regular: TimeInputRange
+  monday: SelectableTimeInputRange
+  tuesday: SelectableTimeInputRange
+  wednesday: SelectableTimeInputRange
+  thursday: SelectableTimeInputRange
+  friday: SelectableTimeInputRange
+}
```

Comment:
Tästä rakenteesta koituu aika paljon copy-paste-henkisyyttä / duplikaatiota. Ite ehkä kokeilisin onnistuisko tyyliin

```
type Regular = { type: 'REGULAR', regular: TimeInputRange }
type Irregular = { type: 'IRREGULAR', monday: ..., tuesday, ..., ... }
type NotSet = { type: 'NOT_SET' }
type FormData = Regular | Irregular | NotSet
```

ja vähentäiskö se ylimääräisiä chekkejä ja duplikaatiota jossain määrin.

Reply:
Lisäksi kaikki `{ monday: T, tuesday: T, ... }` -tyypit on itse asiassa sama kuin `Record<typeof weekdays, T>`, sillä saisi näitä kaikkia tyyppimäärityksiä nätimmäksi, ja varmaan pystyis rakentelemaan apufunktioita jotka luo ja muplaa dataa joka on tässä muodossa.

Reply:
Tuossa ekassa ehdotuksessa ois se drawback, että jos klikkaat radiobuttoneilla eri tyypin jolloin kentät vaihtuu, niin form data ikään kuin resetoituisi koska siinä ei oo mukana kuin valitun tyypin mukaiset kentät, mikä on kyllä sinänsä loogista, mutta jos käyttäjä sitten klikkaa takaisin edellisen valinnan niin kuvittelisin että on ihan kiva jos ne aiemmin syötetyt kellonajat ilmestyy myös takaisin. Siksi lähestyin tätä sillä ajatuksella että tää formData on löyhemmin tyypitetty editoinnin aikana datan säilyttävä rakenne ja ne ylimääräiset kentät vaan piilotetaan. Se on toki totta että kääntöpuolena koodista ei tuu yhtä tiivistä.

Reply:
Totta, hyvä pointti!